feat(server): add subnet_id support

mlinares1998 opened the pull request:
### Summary

This PR adds `subnet_id` support to the `hcloud_server` resource's inline `network` blocks, allowing users to specify which subnet within a network to attach the server to. This brings feature parity with the standalone `hcloud_server_network` resource (PR #1217).

Fixes #1265, #1224

mlinares1998 committed:
feat(server): add subnet_id support

Adds subnet_id field to the inline network block of hcloud_server resource, mirroring functionality from hcloud_server_network.

Fixes #1265, #1224

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

jooola commented on docs/resources/server.md:
````diff
@@ -94,6 +94,27 @@ resource "hcloud_server" "server" {
+    hcloud_network_subnet.network-subnet
+  ]
+}
+```
````
Could you please make this example the recommended one? Bring it on the top and remove the `depends_one` attribute.

Also we probably want to require oneOf network ID or subnet ID, and not allow both.

mlinares1998 replied:
Hi! @jooola 
Sounds great!
I have updated the PR implementing this.

jooola commented on docs/resources/server.md:
```diff
@@ -228,7 +250,8 @@ The following attributes are exported:
 
-- `network_id` - (Required, int) ID of the network
+- `network_id` - (Required, int) ID of the Network to attach the Server to. If `subnet_id` or `ip` are not set, the Server will be attached to the last subnet (ordered by `ip_range`).
+- `subnet_id` - (Optional, string) ID of the Subnet to attach the Server to. When specified, the server will be attached to this specific subnet within the network. The subnet must belong to the network specified in `network_id`.
```
Please require one of networkID or subnetID, since the subnet ID is composed of the network ID and the subnet ip range.

Only having a dependency on the subnet ID makes it simpler for the users.

jooola reviewed (commented):
Hey, thanks for the fix.

mlinares1998 committed:
docs(server): recommend subnet_id for network attachment

mlinares1998 committed:
feat(server): add oneOf validation for network_id and subnet_id

Implement mutually exclusive validation for network attachment fields. Users must now specify exactly one of network_id or subnet_id when attaching servers to networks.

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 reviewed (commented):
Hi @jooola I have implemented the requested changes, could you review the PR again? 😄 
Regards!

mlinares1998 committed:
docs(server): fix typos in network attachment example

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 commented:
Hi @jooola,
I’ve merged the latest changes from main.
Let me know if you have any further suggestions or fixes. :smile: 

Thanks!

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 commented:
Hi @jooola,

I’m following up on this PR as it’s been a few months since the last review.

This change is a critical dependency for hcloud-k8s/terraform-hcloud-kubernetes#163. Currently, the lack of subnet_id support forces us to hardcode node pool IP addresses. This lacks the necessary flexibility and, more importantly, causes unreliable behavior during node pool deletions.

We can't safely resolve these lifecycle issues in the Talos module without this provider-level support.

I'm happy to make any further adjustments if needed. 
Thanks!

M4t7e commented:
Hey @jooola, @apricote,

sorry for the noise, but do you think this can be finished within Q1? We currently have a few issues to solve, and we need this feature in order to change the subnetting layout in [hcloud-k8s/terraform-hcloud-kubernetes](https://github.com/hcloud-k8s/terraform-hcloud-kubernetes), which is a prerequisite for solving the issues properly. If it can't, we'll need to replan and likely postpone everything to the following major release (v5), since changing the subnetting layout will be a breaking change that requires user action and we have to release the next major version (v4) for other reasons latest end of Q1.

I don't want to bug you, but I'd appreciate an indication of whether it's worth waiting for this or not :slightly_smiling_face:

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

marcwickenden commented:
Hi @jooola and @apricote - any update on this? I'm waiting on this fix so terraform-hcloud-kubernetes can be updated and I can migrate a critical workload onto Hetzner.

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

jooola commented:
Hey all,
We are currently really busy with other topics, sorry about that.

I'll try to look at this tomorrow. I can't promise anything though.

jooola committed:
docs: no examples for deprecated feature and move changes to template file

jooola committed:
refactor: use explicit code

laurentpellegrino commented:
@jooola Kindly ping :heart:

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

mlinares1998 commented:
> Hey all, We are currently really busy with other topics, sorry about that.
> 
> I'll try to look at this tomorrow. I can't promise anything though.

Hi! @jooola, i have seen you made changes to the branch.
Anything else we could help you to have this merged soon?
Kindly ping about this :heart:

jooola commented on internal/server/resource.go:
```diff
@@ -1443,11 +1513,29 @@ func validateUniqueNetworkIDs(d *schema.ResourceDiff) error {
-			networkIDI, ok := network["network_id"]
-			if !ok {
-				continue
+			networkID := util.CastInt64(network["network_id"])
```
Could we be a bit more defensive here?

Before `network_id` was required within the `network` block, but not anymore.

I haven't check if using `util.CastInt64` against an undefined network id is safe.

mlinares1998 replied:
Working on it.

jooola commented on internal/server/resource.go:
```diff
@@ -1275,13 +1295,41 @@ func getServerAttributes(d *schema.ResourceData, s *hcloud.Server, forceSetNetwo
+		if nwSet, ok := d.GetOk("network"); ok {
+			for _, item := range nwSet.(*schema.Set).List() {
+				nwData := item.(map[string]interface{})
+				configNetworkID := util.CastInt64(nwData["network_id"])
```
Could we be a bit more defensive when casting the now optional `network_id` attribute? See other comment about similar case.

mlinares1998 replied:
Working on it.

jooola commented on internal/server/resource_test.go:
```diff
@@ -699,6 +699,319 @@ func TestAccServerResource_DirectAttachToNetwork(t *testing.T) {
+					"testdata/r/hcloud_network", nwRes,
+					"testdata/r/hcloud_network_subnet", snwRes,
+					"testdata/r/hcloud_server", sRes,
+				),
```
Was the goal here to test the migration from using `network_id` to using `subnet_id`?

If yes, I am not sure this test is covering this use case, and I think we should definitely test the migration.

Otherwise, it is unclear to me the purpose of this test. Could you help me out a bit?

mlinares1998 replied:
Indeed, will create a separate test for migration.

jooola commented on internal/server/resource_test.go:
```diff
@@ -699,6 +699,319 @@ func TestAccServerResource_DirectAttachToNetwork(t *testing.T) {
+					"testdata/r/hcloud_network_subnet", snw2Res,
+					"testdata/r/hcloud_server", sRes, // Same config with only subnet_id
+				),
+				PlanOnly: true,
```
I recall that the terraform acceptance test framework is checking for this already at each step. So this might be unneeded. But please double check, I might be wrong.

mlinares1998 replied:
Seems to be the case, i'll remove this part.

jooola reviewed (commented):
Here another pass of review. Overrall looks good, but some things could be improved.

Sorry for taking so long to review your PR. The server resource is the most complex resource we have, and we are a bit worried of potentially breaking other users infrastructure. So do not hesitate to add more tests.

jooola commented on internal/server/resource.go:
```diff
@@ -204,8 +204,15 @@ func Resource() *schema.Resource {
+							Type:       schema.TypeInt,
+							Optional:   true,
+							Computed:   true,
+							Deprecated: "Use subnet_id instead. network_id relies on implicit subnet selection which can lead to unpredictable behavior.",
```
While we do recommend using the subnet_id attribute, I am unsure about deprecating the network_id attribute. At least if we deprecate it, we should also deprecate it in the `server_network` resource.

But that should be for another PR.

mlinares1998 replied:
Agree, i'll remove the deprecated flag then. 😄

mlinares1998 reviewed (commented):
@jooola I have improved the PR with your suggestions and added more E2E tests.
Could you give me your feedback 😄 ?

mlinares1998 committed:
refactor(server): defer network_id deprecation, guard network_id casting, add subnet_id E2E tests

mlinares1998 committed:
Merge branch 'main' into feature/server-subnet-id-support

jooola committed:
Merge branch 'main' into feature/server-subnet-id-support

jooola commented:
Thanks for your contribution and your patience!

I'll not release this just yet, and let the e2e tests run a bunch of times on the main branch. I'll try to cut a release #1400 next week if not sooner.
